### ISSUE TITLE
Handle ODK data

### ISSUE THREAD

**Issue by dave-mills** [stats4sd/tape-data-system#6]
The ODK data should be handled and passed through the following process:

1. 'raw' data is brought into the platform via the ODK API and is stored in the database as `submissions.content` (other `submission` columns are metadata for the submission entry). 
2. The data is "processed":
   - Default processing should store all the submission content as new `entities` within the `datasets` linked to the form. For example (from the soils platform), in the screenshot, the main survey is linked to Farmer Field entries, and the crop_repeat_group is linked to the nutrient balance entries. So one submission should create 1 entity within the farmer fields dataset, and 1 entity per repeat within the nutrient balance entries. 
  
![CleanShot 2024-05-01 at 10 45 44](https://github.com/stats4sd/tape-data-system/assets/5711101/09dc0c51-26e6-4c86-ab5f-4c5b8e592911)

**1** is currently working and tested within the Filament ODK Link package, so here we just need to confirm it's working on the TAPE platform. **2** still requires testing, and may require some updates to the odk link package.

**Comment by dave-mills** [comment]
@dan-tang-ssd—Something to note: all the datasets are linked to a particular data model in the platform. For example, the "Farms" dataset is linked to the Farm data model, and the data that is used to generate the CSV lookup tables comes from the `farms` table, not the entities table. 

When handling the data from ODK, I would like us to link the datasets to "real" data models. The models inside `app\Models\SurveyData` are for that purpose. So when a submission comes in, the system would:

- look at the datasets linked to the ODK form and create the appropriate entities and entity_values.
- *and* check if the datasets are linked to specific data models; if so, create the entries there based on. I guess based on some data map. 

We will need to work out how to do that! Perhaps we have a trait that these models can use. Maybe we can set up some sort of standard approach for mapping variables. I've been working on standardising the models that are "lookup tables" - See the `LookupListEntry` Interface and the generic `LookupEntry` class that all the `LookupTables` models extend. I added a `getCsvContentsForOdk()` method that lets the developer specify the format and data for the lookup CSV files. Perhaps we can do something similar for models that get updated by ODK submissions? 

I want to retain this ability to link the `datasets` to real data models and database tables because it's easier for developers to have real data models when developing systems. But the `datasets` and `entities` concepts are still important as they let us be flexible and always fall back on the generic solution whenever there isn't a developed solution.

**Comment by dan-tang-ssd** [comment]
Thank you Dave for the note. 
I will take a look for related program files and try to understand them.

So when we retrieve a submission, it will be stored in:
1. entities and entity_values tables
2. real data model tables (if specified)

Both approach have it's own benefit on different use cases:
E.g. For generic approach, it is good to have a single table to store any entity from any ODK submission
E.g. For real data model approach, it is much quicker to generate CSV lookup tables from real data model tables instead of generic table entities

When we edit a submission, we need to update entites and entity_value tables, we may also need to update the real data model table. (If specified)

**Comment by dan-tang-ssd** [comment]
To test the baseline data retrieval process, I will do below items:
  - ODK central, find and download testing ODK form for tape data system (submitted by Dave before)
  - .env file, check wherher connection to ODK central staging env is setup
  - admin panel, login as admin, create a new Xlsform Template, mark it available for all teams
  - ODK Central, check whether Xlsform Template is submitted
  - admin panel, assign Xlsform Template to the team of "test@example.com"
  - ODK Collect, scan QR code to get ODK form
  - ODK Collect, fill in and submit ODK form
  - ODK Central, check whether ODK submission has been received from Android device
  - app panel, retrive ODK submission from ODK central
  - check whether ODK submission is stored in entities and entity_values tables

P.S. I will study "real" data models related programs after that.

**Comment by dan-tang-ssd** [comment]
I got error in app panel ODK Form Management page. It throws error "Attempt to read property "odkProject" on null".

I did below items but problem persists for both admin account and test account:
1. .env file, remove value for config item "ODK_PLATFORM_PROJECT_ID"
2. php artisan migrate:fresh --seed

Database records look normal...
Um... any idea?

---

Screen shots:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/af6432fa-83e9-40ec-a2ea-49a87e83351c)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/b9e62d9d-5dd7-4bb3-9fe7-2746fbe66542)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/fd466c7f-e87e-4de5-a0ca-f386e6cad214)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/f6376886-d000-4594-86c3-346ff9794407)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/7b133762-4466-4d07-8087-6c1673380a15)

**Comment by dave-mills** [comment]
> I got error in app panel ODK Form Management page. It throws error "Attempt to read property "odkProject" on null".

Looks like the $getRecord() function in the TeamQrCode blade is null. 

Interestingly, locally my team-qr-code.blade.php looks like this: 

```
@if($getState())
    <div class="flex flex-row">
        <div class="mr-4 text-center basis-1/4 border border-gray-400 rounded-lg p-4 bg-white flex flex-col justify-center space-y-4">
            <div class="mx-auto">{{ QrCode::size(150)->generate($getState()) }}</div>
            <h5 class="">SCAN QR Code in ODK Collect</h5>
        </div>

        <div class="basis-3/4 px-12">
           Your project team has been setup. To link your Android device, install and open <b>ODK Central</b>. When asked for project details, scan the QR code on this page. Your device will be linked and you will have access to the forms listed below. Note that only forms marked as <b>Active</b> will be visible within ODK Collect.
        </div>

    </div>
    @else
    <div class="p-4">
        <h5 class="text-center text-danger-600">ODK Project not found. Is this platform connected to a value ODK Central server?</h5>
    </div>
    @endif
```

I'm on the `datasets-with-models` branch for the filament-odk-link package, and up to date with the Github. Maybe try checking that branch out? (I assumed a submodule update would at least get you to the right commit, even if you were not connected to the branch. Maybe not?)

**Comment by dan-tang-ssd** [comment]
Thank you Dave.

I think I checkout dev branch for submodule manually after running "git submodule update" because git status does not show which branch I am using, so I checkout dev branch explicitly yesterday.

I just run "git submodule update", the QR code blade view file is updated as yours.
It works well now, even git status only shows a commit without showing which branch I am using.

---

Screen shots:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/078cb9aa-0518-49f1-af76-fe6044e00847)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/0f63e0aa-1f8a-4d5e-abb8-386eb251757f)

**Comment by dave-mills** [comment]
Yeah, I do wish git submodule update could put you on the correct branch rather than just the right commit. Would make it much easier for this sort of parallel development. Glad it’s working now.


On 2 May 2024, at 14:52, dan-tang-ssd ***@***.***> wrote:

﻿

Thank you Dave.

I think I checkout dev branch for submodule manually after running "git submodule update" because git status does not show which branch I am using, so I checkout dev branch explicitly yesterday.

I just run "git submodule update", the QR code blade view file is updated as yours.
It works well now, even git status only shows a commit without showing which branch I am using.

________________________________

Screen shots:

image.png (view on web)<https://github.com/stats4sd/tape-data-system/assets/86968034/078cb9aa-0518-49f1-af76-fe6044e00847>

image.png (view on web)<https://github.com/stats4sd/tape-data-system/assets/86968034/0f63e0aa-1f8a-4d5e-abb8-386eb251757f>

**Comment by dan-tang-ssd** [comment]
I logged in as admin account.

In admin panel Xlsform Templates page, I create a new record successfully.

In app panel ODK Form Management page, I click "Add a new ODK form" button, then it showed a message mentioning that this feature is not available now... The same for test account.

Um... How do I add a ODK form to a team?

Or is there any other way to test the submission retrieval workflow between application and ODK central?

---

Screen shots:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/36246ee1-e898-4288-8899-adea839d05cd)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/6ebad469-6bfe-402c-9888-c0f1d34bd6a7)

**Comment by dave-mills** [comment]
> Um... How do I add a ODK form to a team?

Admin panel -> teams. There's a relation manager that lets you do it.

**Comment by dave-mills** [comment]
I've left it as an admin feature for now because I want to require some admin interaction to setup a new form, or a new version of a form. (Just while we're building and refining the system).

**Comment by dan-tang-ssd** [comment]
> > Um... How do I add a ODK form to a team?
> 
> Admin panel -> teams. There's a relation manager that lets you do it.

Got it. 

I tried it after getting latest code from dev branch & run migation with seeder files.
Got error about appUsers...

---

Screen shots:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/bfaa7448-0d5f-4722-80a8-4651080c2159)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/f3432261-7d5f-4819-8e1d-87b0ff66d123)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/cab68b75-c06e-4b31-b272-29407b6802eb)

**Comment by dan-tang-ssd** [comment]
It seems like caused by odkProject is null.

I dump $this, which is class Stats4sd\FilamentOdkLink\Models\TeamManagement\Team.
I check odk_projects record for Team, the owner_type is App\Models\Team.

Question:
1. When should we use model class Stats4sd\FilamentOdkLink\Models\TeamManagement\Team?
2. When should we use model class App\Models\Team? 

---

Screen shots:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/3fa15f8b-d689-4b4a-ad62-9bbd7860a251)

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/e24719a3-500a-4457-a86a-afed44520447)

**Comment by dave-mills** [comment]
> When should we use model class Stats4sd\FilamentOdkLink\Models\TeamManagement\Team?
> When should we use model class App\Models\Team?

Within the app, we should always use the App\Models\Team. Within the package, we should probably defer to the config - `models.team_model`. That way, we can defer to the specific app config to know what 'Team' model to use. 

*EDIT*: Does the odk project get correctly setup if you refresh the database? Or is there something specifically different about my setup that I haven't realised and shared? (The 2nd is certainly possible!)

**Comment by dave-mills** [comment]
@dan-tang-ssd  - take a look at #19. I realise that in the .env.example I had strings for the ODK_TEAM_MODEL and ODK_USER_MODEL that didn't work (or make sense). Those are fixed in that PR. I think it's relevant here, as if you have the ODK_TEAM_MODEL set in your local env, the package *should* start using the App's team model everywhere.

**Comment by dan-tang-ssd** [comment]
Status update:

 - admin panel, Xlsform Templates page, create new record
   - ODK central, found ODK form with ID 838
   
 - admin panel, Teams view page, Xlsform relation manager, Add Xlsform to Team
  
 - admin panel, Teams view page, Xlsform relation manager, Publish form
   - Status changed to "LIVE"
   - ODK central, found ODK form with ID 839

 - app panel, ODK Form Management
   - it shows QR code
   - it shows the added Xlsform

---

 - ODK Collect, scan QR code to download Xlsform
   - must select a farm before finalising a form for submission

 - Enketo
   - must select a farm before finalising a form for submission

---

Next step:
- Resolve farm CSV file of Xlsform Template

---

Screen shot:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/f3425f41-a22b-4a22-87ed-512304df5428)

**Comment by dan-tang-ssd** [comment]
I changed to use a simple ODK form (with only one question) for testing.

I added a new table and new dataset (map to the new table), set it to Xlsform Template.
I can submit the ODK form via ODK Collect successfully.

When I retrieve submission in application, submission record created but it seems something wrong when storing data into entities and entity_value tables.

This is good to make the original workflow working properly first.
I will take a deeper look on it.

---

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/8585759d-f51f-4b4a-8860-13916675ad80)

**Comment by dan-tang-ssd** [comment]
I compared table entities in tape data system and AKDN Qol application.
entities.parent_id column is nullable in AKDN, but it is not nullable in Tape.

I changed this column to nullable in TablePlus, and copy class App\Http\Controllers\SubmissionController.php.
Now it can pull submission of a simple ODK form and store it in tables submissions, entities, entity_values successfully.

I will tidy up and push latest code to a new branch. Probably I will use the simple ODK form for studying how to store data to a dedicated table.

---

Screen shot:

![image](https://github.com/stats4sd/tape-data-system/assets/86968034/6012af00-f52f-4fb5-befc-c41265c7ae04)
